nixos/modules/services/monitoring/uptime-kuma.nix

```diff
@@ -0,0 +1,108 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.uptime-kuma;
+in
+{
+
+  options = {
+    services.uptime-kuma = {
+      enable = mkEnableOption (mdDoc "Uptime Kuma");
+
+      dataDir = mkOption {
+        type = types.str;
+        default = "/var/lib/uptimekuma/";
+        description = mdDoc "The directory where Uptime Kuma stores its data files.";
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "uptimekuma";
+        description = mdDoc "User account under which Uptime Kuma runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "uptimekuma";
+        description = mdDoc "Group under which Uptime Kuma runs.";
+      };
+
+      package = mkOption {
+        type = types.package;
+        example = literalExpression "pkgs.uptime-kuma";
+        default = pkgs.uptime-kuma;
+        defaultText = "pkgs.uptime-kuma";
+        description = lib.mdDoc "Uptime Kuma package to use.";
+      };
+
+      settings = lib.mkOption {
+        type =
+          lib.types.submodule { freeformType = with lib.types; attrsOf str; };
+        default = { };
+        description = lib.mdDoc ''
+          Additional configuration for Uptime Kuma, see
+          <https://github.com/louislam/uptime-kuma/wiki/Environment-Variables">
+          for supported values.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    users.users = mkIf (cfg.user == "uptimekuma") {
+      uptimekuma = {
+        isSystemUser = true;
+        group = cfg.group;
+      };
+    };
+
+    users.groups = mkIf (cfg.group == "uptimekuma") { uptimekuma = { }; };
```

Comment:
Unfortunately this is not possible : the software will try to copy kuma.db when launched from it's sources. Because it is launched from the nix-store, it will copy it with read-only permissions and the be stuck. 
As a fix, I am copying it with the right permissions in the pre-start, but we need the state directory to exist at this point, so I have to create it in tmpdir. Tell me if you see a better way to handle the situation.

Reply:
Your `preStart` would work the exact same if run as `DynamicUser`. What am I missing?

Reply:
I need to create the folder before the preStart, then how do I create the folder now knowing the user ? (Maybe _I_ am missing something).

Reply:
`systemd.services.uptime-kuma.serviceConfig.StateDirectory = "uptime-kuma";` will provision a directory for you which you can then write into for persistent storage.

Reply:
Yup, but when I tried that I found out that this directory was created after the `preStart`.

Reply:
I'll give this some thought.

Reply:
@RaitoBezarius do you have any feedback here?

Reply:
@aanderse this seems to be completely https://github.com/systemd/systemd/issues/16060

Reply:
which seems to suggest we should use `LoadCredential` here, but it would only give read only access IMHO

Reply:
Plus:

> Hmm, not sure I follow? This is simply about adding a way how services can be reset to their factory defaults, simply by flushing out their state data. Of course, services shouldn't using StateDirectory= if they can't start with an empty state directory in the first place...

So really, this is more like a bug in the program which should copy its own state from its template in its data directory, this hack seems acceptable to me, in the meantime.

Reply:
Thanks for looking.

Reply:
Hi @aanderse @RaitoBezarius, thinking back about this question and reading a little bit more the code of the upstream project, I've decided to open a PR upstream to have the database copied with the right permissions in the StateDirectory and be able to write a more clean module (see [here](https://github.com/louislam/uptime-kuma/pull/2119)). For now, I've rewritten the module and included a patch, and I'd be happy if we could merge the PR with the patch for now. When/if the upstream PR is included in a release, I'll simply remove the patch. What do you think about this solution ?